pkg/planner/core/operator/physicalop/physical_utils.go

```diff
@@ -159,3 +166,128 @@ func ClonePhysicalPlansForPlanCache(newCtx base.PlanContext, plans []base.Physic
 	}
 	return clonedPlans, true
 }
+
+// EncodeUniqueIndexKey encodes a unique index key.
+func EncodeUniqueIndexKey(ctx sessionctx.Context, tblInfo *model.TableInfo, idxInfo *model.IndexInfo, idxVals []types.Datum, tID int64) (_ []byte, err error) {
+	encodedIdxVals, err := EncodeUniqueIndexValuesForKey(ctx, tblInfo, idxInfo, idxVals)
+	if err != nil {
+		return nil, err
+	}
+	return tablecodec.EncodeIndexSeekKey(tID, idxInfo.ID, encodedIdxVals), nil
+}
+
+// EncodeUniqueIndexValuesForKey encodes unique index values for a key.
+func EncodeUniqueIndexValuesForKey(ctx sessionctx.Context, tblInfo *model.TableInfo, idxInfo *model.IndexInfo, idxVals []types.Datum) (_ []byte, err error) {
+	sc := ctx.GetSessionVars().StmtCtx
+	for i := range idxVals {
+		colInfo := tblInfo.Columns[idxInfo.Columns[i].Offset]
+		// table.CastValue will append 0x0 if the string value's length is smaller than the BINARY column's length.
+		// So we don't use CastValue for string value for now.
+		// TODO: The first if branch should have been removed, because the functionality of set the collation of the datum
+		// have been moved to util/ranger (normal path) and getNameValuePairs/getPointGetValue (fast path). But this change
+		// will be cherry-picked to a hotfix, so we choose to be a bit conservative and keep this for now.
+		if colInfo.GetType() == mysql.TypeString || colInfo.GetType() == mysql.TypeVarString || colInfo.GetType() == mysql.TypeVarchar {
+			var str string
+			str, err = idxVals[i].ToString()
+			idxVals[i].SetString(str, idxVals[i].Collation())
+		} else if colInfo.GetType() == mysql.TypeEnum && (idxVals[i].Kind() == types.KindString || idxVals[i].Kind() == types.KindBytes || idxVals[i].Kind() == types.KindBinaryLiteral) {
+			var str string
+			var e types.Enum
+			str, err = idxVals[i].ToString()
+			if err != nil {
+				return nil, kv.ErrNotExist
+			}
+			e, err = types.ParseEnumName(colInfo.FieldType.GetElems(), str, colInfo.FieldType.GetCollate())
+			if err != nil {
+				return nil, kv.ErrNotExist
+			}
+			idxVals[i].SetMysqlEnum(e, colInfo.FieldType.GetCollate())
+		} else {
+			// If a truncated error or an overflow error is thrown when converting the type of `idxVal[i]` to
+			// the type of `colInfo`, the `idxVal` does not exist in the `idxInfo` for sure.
+			idxVals[i], err = table.CastValue(ctx, idxVals[i], colInfo, true, false)
+			if types.ErrOverflow.Equal(err) || types.ErrDataTooLong.Equal(err) ||
+				types.ErrTruncated.Equal(err) || types.ErrTruncatedWrongVal.Equal(err) {
+				return nil, kv.ErrNotExist
+			}
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	encodedIdxVals, err := codec.EncodeKey(sc.TimeZone(), nil, idxVals...)
+	err = sc.HandleError(err)
+	if err != nil {
+		return nil, err
+	}
+	return encodedIdxVals, nil
+}
+
+// CloneConstant2DForPlanCache clones a 2D slice of *expression.Constant.
+func CloneConstant2DForPlanCache(constants [][]*expression.Constant) [][]*expression.Constant {
```

Comment:
can we add a todo: here, since it's about plan cache, we should move them in the future with all following func
```
cloneExpressionsForPlanCache
cloneExpression2DForPlanCache
cloneColumnsForPlanCache
...
plan_clone_utils.go
```